### ISSUE TITLE
Faster_R_CNN

### ISSUE THREAD

**Issue by yushenxiang** [yushenxiang/MachineLearningTips#1]
代码链接：
https://github.com/rbgirshick/py-faster-rcnn

论文阅读总结：
1）提出一个和目标检测网络共享全部卷积层的RPN网络来加快region proposals的速度，解决以往fast-r-cnn和sppnet的region proposals计算的瓶颈问题
2）RPN网络在fast-r-cnn中是用来进行目标检测的，作者使用attenion机制通过共享全部卷积层来将RPN和fast-r-cnn整合到一起。
3）实现效果：vgg-16网络上，实现在单个gpu上5fps ，以及MS COCO数据集上每张图300个region proposals

介绍：
1）以往方法：
Selective Search： at 2 seconds per image in a CPU implementation.
EdgeBoxes：at 0.2 seconds per image.
2） Region Proposal Networks (RPNs)：
share convolutional layers with state-of-the-art object detection networks . By sharing convolutions at test-time, the marginal cost for computing proposals is small （ 10ms per image)
RPN实现原理：
Our observation is that the convolutional feature maps used by region-based detectors, like Fast R- CNN, can also be used for generating region proposals. On top of these convolutional features, we construct an RPN by adding a few additional convolutional layers that simultaneously regress region bounds and objectness scores at each location on a regular grid.
RPN实现region proposals预测的方法：
instead of pyramids of images and pyramids of filters, we introduce novel “anchor” boxes that serve as references at multiple scales and aspect ratios. Our scheme can be thought of as a pyramid of regression references , which avoids enumerating images or filters of multiple scales or aspect ratios. RPN与fast-r-cnn检测网络同时工作的方法：
we propose a training scheme that alternates between fine-tuning for the region proposal task and then fine-tuning for object detection, while keeping the proposals fixed.

2)相关研究工作：
object proposals（物体候选）：
Widely used object proposal methods include those based on grouping super-pixels (e.g., Selective Search , CPMC ， MCG ) and those based on sliding windows (e.g., objectness in windows ， EdgeBoxes ）

Deep Networks for Object Detection:
R-CNN 👍 
R-CNN mainly plays as a classifier( to classify the proposal regions into object categories or background.), and it does not predict object bounds (except for refining by bounding box regression). Its accuracy depends on the perfor- mance of the region proposal module 
The OverFeat:
computes convolutional features from an image pyramid for classification, localization, and de- tection. Adaptively-sized pooling (SPP) 👍 
on shared convolutional feature maps is developed for efficient region-based object detection [1], [30] and semantic segmentation [29]. 
Fast R-CNN:
 enables end-to-end detector training on shared convolutional features and shows compelling accuracy and speed.

3)FASTER R-CNN:
The first module is a deep fully convolutional network that proposes regions, 
and the second module is the Fast R-CNN detector that uses the proposed regions. 
with ‘attention’ mechanisms, the RPN module tells the Fast R-CNN module where to look.

**3.1 Region Proposal Networks**
A Region Proposal Network (RPN) takes an image (of any size) as input and outputs a set of rectangular object proposals, each with an objectness score.
具体实现过程：
To generate region proposals, we slide a small network over the convolutional feature map output by the last shared convolutional layer. 
This small network takes as input an 3x3 spatial window of the input convolutional feature map. Each sliding window is mapped to a lower-dimensional feature (256-d for ZF and 512-d for VGG, with ReLU following). 
This feature is fed into two sibling fully- connected layers—a box-regression layer (reg) and a box-classification layer (cls). 

3.1.1 Anchors
在每一个sliding window都会有k(这里是9）个region proposals. 这些region proposals是由anchors实现的。
 each anchor is centered at the sliding window in question, and is associated with a scale and aspect ratio 
regression layer输出4k个值计算出所有region proposal的坐标，
classification layer输出2k个值，计算出所有的region proposal到底是物体还是背景

优势：
image/feature pyramids：
需要把图片转换成不同尺寸和比例大小，计算量大
pyramid of filters：
需要不同尺寸和比例大小的sliding window，通常和第一种方法一起使用
a pyramid of anchors：
只需要使用一个尺寸大小的图片和一个固定大小的sliding window或filters，通过对不同尺寸和比例的anchors进行regressiong and classification得到物体框位置和分类。
（这是faster-rcnn的核心优势）

3.1.2 Loss Function
训练RPN：
每一个anchor都会被赋予两个值，分别代表是物体或不是物体；
1）根据IOU来给anchor判定是否是正样本：
 (i) 和一个ground truth box有最大的IOU的anchor, （通常采用这个方法）
or (ii) 只要和任意一个ground-truth box的IOU大与0.7就可以
2）根据IOU来给anchor判定是否是负样本：
和所有的ground truth box的IOU都小于0.3
3）而那些非正非负样本的anchors都对训练目标没有影响。

loss函数的输出：
anchor是物体的概率和目标限制框的4个坐标值

目标限制框的定位:
相当于从anchor box尽量回归接近附近的ground truth box
没有使用ROI（Region of Interest）的方法:
ROI是根据一系列不同尺寸的ROI区域得到的不同尺寸大小的特征，进行回归计算
Faster-rcnn是使用特征图上一个固定尺寸大小（3x3）的特征进行回归计算，通过不同大小的anchor来计算不同大小的目标限制框。
因此正是由于anchor的引入可以实现在特征保持固定大小不变的情况下准确地道不同大小的目标框的回归。

超参数：
1）由于负样本占据大多数，因此我们每次使用256作为一个批次进行学习（其中正负样本各占一半）；
classification term is normalized by the mini-batch size (i.e., Ncls = 256) 
2）and the reg term is normalized by the number of anchor locations (i.e., Nreg ⇠ 2, 400). 
3）By default we set   lamda= 10, and thus both cls and reg terms are roughly equally weighted. 
4）所有新加的层权重的初始值都采取a zero-mean Gaussian distribution with standard deviation 0.01.
5）所有的共享卷积层采用已经预训练好的网络权重
6） learning rate of 0.001 for 60k mini-batches, and 0.0001 for the next 20k mini-batches on the PASCAL VOC dataset.、
7） We use a momentum of 0.9 and a weight decay of 0.0005

**这里我们已经实现了如何得到图片中包含物体的限制框以及位置，但是还没有对这些物体进行检测判断到底是什么物体（人？车？狗？树？）
下一步将会介绍如何利用RPN和fast-r-cnn的检测网络联合起来得到快速的目标检测。**

**3.2 Sharing Features for RPN and Fast R-CNN**
RPN和Fast R-CNN在独立训练时的卷积层可能是不一样的，因此需要一起训练，来得到共享的卷积层网络。
主要有三种方法：
1）交替训练Alternating training：（faster r-cnn 采用的方法）
先训练RPN，然后利用得到的region proposal训练Fast R-CNN的检测网络。再利用训练的Fast R-CNN的检测网络来初始化RPN,不停迭代这个过程。

具体的4步联合训练过程：

- 训练RPN (This network is initialized with an ImageNet-pre-trained model _

- 使用第一步训练RPN得到的region proposals进行单独的Fast R-CNN的训练
（This detection net- work is also initialized by the ImageNet-pre-trained model.）

- 使用第二步训练的Fast R-CNN网络进行初始化RPN，但是保留共享的卷积层参数，只细调那些RPN独有的层（这时候的训练已经开始共享卷积层了）

- 最后，保持共享的卷积层参数不变，只细调那些Fast R-CNN网络独有的那些层

实验细节：
1）For anchors, we use 3 scales with box areas of 1282, 2562, and 5122 pixels, and 3 aspect ratios of 1:1, 1:2, and 2:1. 
2）the total stride for both ZF and VGG nets on the last convolutional layer is 16 pixels, 
and thus is ⇠10 pixels on PASCAL
3）为了减少region proposal互相重叠带来的影响，根据非最大值抑制法(NMS)，基于region proposal的classification score.将NMS的IOU阈值设置为0.7，那么每张图片将只保留2000个region proposal。NMS的引入在不影响最终检测精度的情况下减少了region proposal的数量。
4）之后，使用top-N排列的region proposal进行目标检测。（文章实验证明哪怕只使用top-300，也能得到很高的检测精度)

2）近似联合训练Approximate joint training：
RPN和Fast R-CNN直接作为一个整体进行训练。在每一次随机梯度下降迭代的过程中，前向传播产生会被fast-r-cnn进行检测训练时用的region proposals,后向传播时会计算RPN和fast-r-cnn总的损失函数。
这个方法很容易实现，但忽略了一点，那就是本应该是网络输出一部份的物体限制框的坐标值被忽略了。
因此被称为近似训练。这种方法能够降低25%甚至50%的训练时间。

3）非参数联合训练Non-approximate joint training：
通过使用一个RoI pooling layer来对物体目标框的坐标进行计算来弥补近似联合训练的不足。

架构备注：
For the ImageNet pre-trained network, we use the “fast” version of ZF net  that has 5 convolutional layers and 3 fully-connected layers, 
and the public VGG-16 mode that has 13 con- volutional layers and 3 fully-connected layers.
